### PR TITLE
Add status_code_and_message interop test.

### LINKED ISSUE #742: Implement status_code_and_message interop test

### PR DESCRIPTION
**Pull request by markdroth**
Fixes #742.

This seems to compile, but I don't know how to test whether it actually works.  Is there documentation somewhere about how to run the interop tests?

### CODE REVIEW

**Comment by markdroth** [comment]
I think I've figured out how to run the interop tests (using the script from the grpc/grpc repo).  However, the new test is failing with the following error:

2016/07/01 15:51:43 /TestService/UnaryCall unexpected RPC error message: rpc error: code = 2 desc = test status message
exit status 1
FAILED: cloud_to_cloud:go:go_server:status_code_and_message [ret=1, pid=4074]
FAILED: Some tests failed

I'm guessing that this is being triggered by my code to return an error from the method handlers on the server side.  I was thinking that would cause the server to return a non-successful status, but it looks like that doesn't work.  What's the right way to have the server return a different status?

**Comment by markdroth** [comment]
Actually, nevermind that last comment -- it looks like that error is actually coming from the new test itself, where err.Error() returns more than just the RPC error message.  Is there a way to extract just the RPC error message from the error object?

File: interop/test_utils.go [hunk 1]
```diff
@@ -454,6 +454,49 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
 	}
 }
 
+// DoStatusCodeAndMessage checks that the status code is propagated back to the client.
+func DoStatusCodeAndMessage(tc testpb.TestServiceClient) {
+	var code int32 = 2
+	msg := "test status message"
+	respStatus := &testpb.EchoStatus{
+		Code:    &code,
+		Message: &msg,
```
Comment (iamqizhao) [inline interop/test_utils.go]: respStatus := &testpb.EchoStatus{
  Code: proto.Int32(2),
  Message: proto.String("test status message",
}

File: interop/test_utils.go [hunk 2]
```diff
@@ -454,6 +454,49 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
...
+	}
+	// Test UnaryCall.
+	req := &testpb.SimpleRequest{
+		ResponseStatus: respStatus,
+	}
+	_, err := tc.UnaryCall(context.Background(), req)
```
Comment (iamqizhao) [inline interop/test_utils.go]: if _, err := tc.UnaryCall(...); grpc.Code(err) != codes.Code(code) {
   ...
}

File: interop/test_utils.go [hunk 3]
```diff
@@ -454,6 +454,49 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
...
+	if grpc.Code(err) != codes.Code(code) {
+		grpclog.Fatalf("/TestService/UnaryCall RPC compleled with error code %d, want %d", grpc.Code(err), code)
+	}
+	if err.Error() != fmt.Sprint("rpc error: code = ", code, " desc = ", msg) {
+		grpclog.Fatal("/TestService/UnaryCall unexpected RPC error message: ", err)
+	}
```
Comment (iamqizhao) [inline interop/test_utils.go]: I think the above can be better represented as:

expectedErr := grpc.Errorf(code, msg)
if _, err := tc.UnaryCall(...); err == nil || err.Error() != expectedErr.Error() {
   grpclog.Fatalf("%v.UnaryCall(_, _) = _, %v, want %v", err, expectedErr)
}

File: interop/test_utils.go [hunk 4]
```diff
@@ -454,6 +454,49 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
...
+	// Test FullDuplexCall.
+	stream, err := tc.FullDuplexCall(context.Background())
+	if err != nil {
+		grpclog.Fatalf("%v.FullDuplexCall(_) = _, %v", tc, err)
```
Comment (iamqizhao) [inline interop/test_utils.go]: grpclog.Fatalf("%v.FullDuplexCall(_) = _, %v, want <nil>", tc, err)

File: interop/test_utils.go [hunk 5]
```diff
@@ -454,6 +454,49 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
...
+	}
+	stream_req := &testpb.StreamingOutputCallRequest{
+		ResponseStatus: respStatus,
+	}
+	if err := stream.Send(stream_req); err != nil {
+		grpclog.Fatalf("stream %v.Send(%v) = %v", stream, stream_req, err)
```
Comment (iamqizhao) [inline interop/test_utils.go]: ditto -- add "want ..."

File: interop/test_utils.go [hunk 6]
```diff
@@ -454,6 +454,49 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
...
+	}
+	err = stream.CloseSend()
```
Comment (iamqizhao) [inline interop/test_utils.go]: if err := stream.CloseSend(); err != nil {
  ...
}

File: interop/test_utils.go [hunk 7]
```diff
@@ -454,6 +454,49 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
...
+	if err != nil {
+		grpclog.Fatalf("stream %v.CloseSend() = %v", stream, err)
+	}
+	_, err = stream.Recv()
```
Comment (iamqizhao) [inline interop/test_utils.go]: same as the unary case.

File: interop/test_utils.go (hunk 1 above)
Comment (markdroth) [inline interop/test_utils.go]: Done.

File: interop/test_utils.go (hunk 2 above)
Comment (markdroth) [inline interop/test_utils.go]: Can't use that here, because I'm using the value of `err` again on line 473, outside of this `if` statement.

But probably no longer relevant, given the change you suggested below.

File: interop/test_utils.go (hunk 3 above)
Comment (markdroth) [inline interop/test_utils.go]: Done.

File: interop/test_utils.go (hunk 4 above)
Comment (markdroth) [inline interop/test_utils.go]: In this case, we don't want any error, so any failure is a problem.

File: interop/test_utils.go (hunk 5 above)
Comment (markdroth) [inline interop/test_utils.go]: We don't want any error here.

File: interop/test_utils.go (hunk 6 above)
Comment (markdroth) [inline interop/test_utils.go]: Done.

File: interop/test_utils.go (hunk 7 above)
Comment (markdroth) [inline interop/test_utils.go]: Done.

**Comment by markdroth** [comment]
Note that I've changed grpc/grpc#7201 to test this against the C++ implementation, which now passes.

File: interop/test_utils.go (hunk 4 above)
Comment (iamqizhao) [inline interop/test_utils.go]: github hid part of my comments. Please take a look at my original comment again. In idiomatic golang test, we typically add "want <nil>" if this function call does not expect any error.

File: interop/test_utils.go (hunk 5 above)
Comment (iamqizhao) [inline interop/test_utils.go]: `want <nil>`

File: interop/test_utils.go (hunk 4 above)
Comment (markdroth) [inline interop/test_utils.go]: Done.

File: interop/test_utils.go (hunk 5 above)
Comment (markdroth) [inline interop/test_utils.go]: Done.

File: interop/test_utils.go [hunk 8]
```diff
@@ -454,6 +454,41 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
...
+	respStatus := &testpb.EchoStatus{
+		Code:    proto.Int32(code),
+		Message: proto.String(msg),
+	}
+	// Test UnaryCall.
+	req := &testpb.SimpleRequest{
+		ResponseStatus: respStatus,
+	}
+	if _, err := tc.UnaryCall(context.Background(), req); err.Error() != expectedErr.Error() {
+		grpclog.Fatalf("UnaryCall RPC returned error %v, want %v", err, expectedErr)
```
Comment (menghanl) [inline interop/test_utils.go]: `grpclog.Fatalf("%v.UnaryCall(_, %v) = _, %v, want _, %v", tc, req, err, expectedErr)`

File: interop/test_utils.go [hunk 9]
```diff
@@ -454,6 +454,41 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
...
+	// Test FullDuplexCall.
+	stream, err := tc.FullDuplexCall(context.Background())
+	if err != nil {
+		grpclog.Fatalf("%v.FullDuplexCall(_) = _, %v, want <nil>", tc, err)
+	}
+	stream_req := &testpb.StreamingOutputCallRequest{
+		ResponseStatus: respStatus,
+	}
+	if err := stream.Send(stream_req); err != nil {
+		grpclog.Fatalf("stream %v.Send(%v) = %v, want <nil>", stream, stream_req, err)
```
Comment (menghanl) [inline interop/test_utils.go]: Remove "stream" from fatalf

File: interop/test_utils.go [hunk 10]
```diff
@@ -454,6 +454,41 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
...
+	}
+	if err := stream.CloseSend(); err != nil {
+		grpclog.Fatalf("stream %v.CloseSend() = %v, want <nil>", stream, err)
```
Comment (menghanl) [inline interop/test_utils.go]: Remove "stream" from fatalf

File: interop/test_utils.go [hunk 11]
```diff
@@ -454,6 +454,41 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
...
+	}
+	if _, err = stream.Recv(); err.Error() != expectedErr.Error() {
+		grpclog.Fatalf("stream %v.Recv() returned error %v, want %v", err, expectedErr)
```
Comment (menghanl) [inline interop/test_utils.go]: Remove "stream" from fatalf. And you missed the variable for the first `%v`
`grpclog.Fatalf("%v.Recv() returned error %v, want %v", stream, err, expectedErr)`

File: interop/test_utils.go [hunk 12]
```diff
@@ -454,6 +454,41 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
 	}
 }
 
+// DoStatusCodeAndMessage checks that the status code is propagated back to the client.
+func DoStatusCodeAndMessage(tc testpb.TestServiceClient) {
+	var code int32 = 2
+	msg := "test status message"
+        expectedErr := grpc.Errorf(codes.Code(code), msg)
```
Comment (menghanl) [inline interop/test_utils.go]: Extra space at the beginning.

File: interop/test_utils.go (hunk 8 above)
Comment (markdroth) [inline interop/test_utils.go]: Done.

File: interop/test_utils.go (hunk 9 above)
Comment (markdroth) [inline interop/test_utils.go]: Done.

File: interop/test_utils.go (hunk 10 above)
Comment (markdroth) [inline interop/test_utils.go]: Done.

File: interop/test_utils.go (hunk 11 above)
Comment (markdroth) [inline interop/test_utils.go]: Done.  (I'm surprised the compiler didn't catch this.)

File: interop/test_utils.go (hunk 12 above)
Comment (markdroth) [inline interop/test_utils.go]: Looks like the tab got expanded into spaces.  Fixed.

**Comment by markdroth** [comment]
Ping?

File: interop/test_utils.go [hunk 13]
```diff
@@ -485,6 +520,10 @@ func serverNewPayload(t testpb.PayloadType, size int32) (*testpb.Payload, error)
 }
 
 func (s *testServer) UnaryCall(ctx context.Context, in *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+	resp := in.GetResponseStatus()
```
Comment (iamqizhao) [inline interop/test_utils.go]: s/resp/status/g? "resp" is misleading.

File: interop/test_utils.go [hunk 14]
```diff
@@ -540,6 +579,10 @@ func (s *testServer) FullDuplexCall(stream testpb.TestService_FullDuplexCallServ
 		if err != nil {
 			return err
 		}
+		resp := in.GetResponseStatus()
```
Comment (iamqizhao) [inline interop/test_utils.go]: why do we call in.GetResponseStatus twice in this func?

**Comment by iamqizhao** [comment]
Can you rebase? If the travis build is still failed, please fix it or let me know.

File: interop/test_utils.go (hunk 13 above)
Comment (markdroth) [inline interop/test_utils.go]: Done.

File: interop/test_utils.go (hunk 14 above)
Comment (markdroth) [inline interop/test_utils.go]: We don't.  It's only being called once.  (The call right below this one is to GetResponseParameters(), not GetResponseStatus().)

**Comment by markdroth** [comment]
Looks like the travis build is still failing, but I don't see any obvious errors in the log, so I'm not sure what's wrong here.

File: interop/test_utils.go [hunk 15]
```diff
@@ -454,6 +454,41 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
...
+	}
+	if _, err := tc.UnaryCall(context.Background(), req); err.Error() != expectedErr.Error() {
+		grpclog.Fatalf("%v.UnaryCall(_, %v) = _, %v, want _, %v", tc, req, err, expectedErr)
+	}
+	// Test FullDuplexCall.
+	stream, err := tc.FullDuplexCall(context.Background())
+	if err != nil {
+		grpclog.Fatalf("%v.FullDuplexCall(_) = _, %v, want <nil>", tc, err)
+	}
+	stream_req := &testpb.StreamingOutputCallRequest{
```
Comment (menghanl) [inline interop/test_utils.go]: go vet is complaining about this line
don't use underscores in Go names; var stream_req should be streamReq

Comment (markdroth) [inline interop/test_utils.go]: Fixed.

File: interop/test_utils.go [hunk 16]
```diff
@@ -454,6 +454,41 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
 	}
 }
 
+// DoStatusCodeAndMessage checks that the status code is propagated back to the client.
+func DoStatusCodeAndMessage(tc testpb.TestServiceClient) {
+	var code int32 = 2
```
Comment (iamqizhao) [inline interop/test_utils.go]: can we simply use codes.Unknown directly in line#461 and 463?

File: interop/test_utils.go [hunk 17]
```diff
@@ -454,6 +454,41 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
...
+	expectedErr := grpc.Errorf(codes.Code(code), msg)
+	respStatus := &testpb.EchoStatus{
+		Code:    proto.Int32(code),
+		Message: proto.String(msg),
+	}
+	// Test UnaryCall.
+	req := &testpb.SimpleRequest{
+		ResponseStatus: respStatus,
+	}
+	if _, err := tc.UnaryCall(context.Background(), req); err.Error() != expectedErr.Error() {
```
Comment (iamqizhao) [inline interop/test_utils.go]: `if _, err := ....; err == nil || err.Error() != expectedError() {`

Comment (iamqizhao) [inline interop/test_utils.go]: perhaps add grpc.FailFast(false) as the call option to prevent network hiccup fails the test?

File: interop/test_utils.go [hunk 18]
```diff
@@ -454,6 +454,41 @@ func DoCancelAfterFirstResponse(tc testpb.TestServiceClient) {
...
+		grpclog.Fatalf("%v.UnaryCall(_, %v) = _, %v, want _, %v", tc, req, err, expectedErr)
+	}
+	// Test FullDuplexCall.
+	stream, err := tc.FullDuplexCall(context.Background())
```
Comment (iamqizhao) [inline interop/test_utils.go]: set grpc.FailFast(false)?

File: interop/test_utils.go (hunk 16 above)
Comment (markdroth) [inline interop/test_utils.go]: I could do that on line 461.  However, I can't seem to make it work on line 463; when I try, the compiler says this:

interop/test_utils.go:463: cannot use codes.Unknown (type codes.Code) as type *int32 in field value

Since I'd like both lines to use the same value, it seemed better to define an int32 and use it in both places.

File: interop/test_utils.go (hunk 17 above)
Comment (markdroth) [inline interop/test_utils.go]: Done.

Comment (markdroth) [inline interop/test_utils.go]: None of the other interop tests seem to use fail-fast, so let's keep things consistent.  (If we want to change all of the tests to use fail-fast, we can do that in a separate PR.)

File: interop/test_utils.go (hunk 18 above)
Comment (markdroth) [inline interop/test_utils.go]: See above.

**Comment by markdroth** [comment]
Still waiting for an approval on this.
